### DISCUSSION TITLE
Why directx ?

### DISCUSSION THREAD

**Discussion by jackffmm** [godotengine/godot-proposals#7644] [Engine Core]
Why godots use directx on windows ? vulkan more than enaught.
You waste resources!
Also support opengl is waste.

**Comment by Calinou** [comment]
Godot currently doesn't use Direct3D – this is being implemented by https://github.com/godotengine/godot/pull/70315#. When that pull request is merged, the ability to use Vulkan and OpenGL on Windows will remain.

There are a few good reasons to use Direct3D on Windows instead of Vulkan or OpenGL:

- It's the API Microsoft cares about the most. This means various automatic Windows optimizations (such as decreasing input lag) are only effective on Direct3D applications, not Vulkan or OpenGL.
- Intel's Windows driver has relatively poor Vulkan and OpenGL support compared to Direct3D.
- AMD's Windows driver has relatively poor OpenGL support compared to Direct3D. Performance is often worse compared to NVIDIA on hardware that exhibits equivalent performance in Direct3D applications.

Some of the benefits of using Direct3D can be used with Vulkan by [presenting with DXGI](https://github.com/godotengine/godot-proposals/issues/5692), which is an hybrid form where Direct3D is used for presentation and Vulkan is used for rendering. This doesn't sidestep the issues with Intel's Vulkan drivers on Windows though.

> You waste resources!

The Direct3D 12 driver reuses a lot of existing code. Core shaders don't need to be rewritten, as they're translated from GLSL to HLSL automatically. User-provided shaders continue to work as usual using the same approach.

> Also support opengl is waste.

Lots of hardware used by people in the Godot community doesn't support Vulkan yet, so we'll keep maintaining the OpenGL renderer for the foreseeable future. This is especially the case on mobile and web platforms (the latter can't run Vulkan in any form, and WebGPU is not a 1:1 replacement).

**Reply by zocker-160** [reply]
so much for https://github.com/godotengine/godot/issues/23166#issuecomment-431560703
> Godot is cross-platform at its core, and won't bother supporting a proprietary API for a single platform if there are open source alternatives (OpenGL, Vulkan)

> It's the API Microsoft cares about the most
> Intel's Windows driver
> AMD's Windows driver

If you care so much about bad proprietary GPU drivers on Windows, why did you even bother with Vulkan in the first place? 
Is now anything except Windows a second class citizen in Godot?

**Comment by Jesusemora** [comment]
Console support. People asked for console support so the godot devs started working on it.
I believe XBox is one that only supports directx.
I imagine directx will be optional just as vulkan and opengl are right now.

**Reply by zocker-160** [reply]
> Console support. People asked for console support so the godot devs started working on it.

Godot cannot support any console officially.

Quote from the PR linked above

> XBox —which can't be supported officially by Godot

Ironically the most relevant consoles (PS5 and Switch) both don't support DX.

Them adding support for something that cannot be officially supported? Huh?

>  I imagine directx will be optional just as vulkan and opengl are right now.

The issue is that if the argument for DX support is "Windows market share" then I fear that OpenGL and Vulkan backends are going to degrade in quality and performance compared to DX, since by that same argument DX is the only relevant backend.

**Reply by AThousandShips** [reply]
> The issue is that if the argument for DX support is "Windows market share" then I fear that OpenGL and Vulkan backends are going to degrade in quality and performance compared to DX, since by that same argument DX is the only relevant backend.

Based on what? It's an [*alternative* backend](https://docs.godotengine.org/en/latest/about/faq.html#why-does-godot-prioritize-vulkan-and-opengl-over-direct3d), not the primary one, the primary one is and remains Vulkan, which gets far more attention

> since by that same argument DX is the only relevant backend.

An *optional* backend for *one* OS will not make everyone just suddenly ignore the main backend for all platforms but one

**Reply by zocker-160** [reply]
> Based on what?

Based on your own argument, that Windows market share and bad Windows drivers are the reason for DX support in the first place. Any additional backend will cost time and resources that are not spent on making Vulkan better.

> the primary one is and remains Vulkan

This statement will probably age as well as the first bullet point here: https://github.com/godotengine/godot/issues/23166#issuecomment-431560703

> An optional backend for one OS will not make everyone just suddenly ignore the main backend for all platforms but one

You are right, not suddenly, but gradually over time.

**Reply by AThousandShips** [reply]
> Any additional backend will cost time and resources that are not spent on making Vulkan better.

I don't necessarily disagree with this

But that's very different from the completely unfounded assumption that direct3D will just somehow take over, going against not just basic reason but the active intentions of the maintainers of the rendering systems as well as general priorities

**Reply by bruvzg** [reply]
> https://github.com/godotengine/godot/issues/23166#issuecomment-431560703

It is a bit different situation. DirectX 9 was a completely different API from existing OpenGL and require fully independent renderer. DirectX 12 is quite similar to Vulkan and Metal (which is also in development - https://github.com/godotengine/godot/pull/88199), and a lot of the renderer code is shared as common `RendererRD`. There's not much to take over.

Shared part - https://github.com/godotengine/godot/tree/master/servers/rendering/renderer_rd
Vulkan part - https://github.com/godotengine/godot/tree/master/drivers/vulkan
D3D12 part - https://github.com/godotengine/godot/tree/master/drivers/d3d12

Also, for the case of bad drivers Godot is using DX11 as well (as fallback for OpenGL via ANGLE translation layer, there's no separate renderer).

**Comment by catboy-catfish** [comment]
> Why godots use directx on windows ? vulkan more than enaught.
> 
> You waste resources!
> 
> Also support opengl is waste.

Vulkan may be faster than DirectX or OpenGL, but it's not as stable as them right now due to it being more recent.
